src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java

```diff
@@ -0,0 +1,480 @@
+package org.zalando.nakadi.service.subscription.zk;
+
+import com.google.common.base.Charsets;
+import static com.google.common.base.Charsets.UTF_8;
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.NodeCache;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+import org.echocat.jomon.runtime.concurrent.RetryForSpecifiedCountStrategy;
+import static org.echocat.jomon.runtime.concurrent.Retryer.executeWithRetry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.zalando.nakadi.domain.EventTypePartition;
+import org.zalando.nakadi.exceptions.NakadiRuntimeException;
+import org.zalando.nakadi.exceptions.ServiceUnavailableException;
+import org.zalando.nakadi.exceptions.UnableProcessException;
+import org.zalando.nakadi.exceptions.runtime.OperationInterruptedException;
+import org.zalando.nakadi.exceptions.runtime.OperationTimeoutException;
+import org.zalando.nakadi.exceptions.runtime.RequestInProgressException;
+import org.zalando.nakadi.exceptions.runtime.ZookeeperException;
+import org.zalando.nakadi.service.subscription.model.Session;
+import org.zalando.nakadi.view.SubscriptionCursorWithoutToken;
+
+public abstract class AbstractZkSubscriptionClient implements ZkSubscriptionClient {
+    private static final String STATE_INITIALIZED = "INITIALIZED";
+    private static final int COMMIT_CONFLICT_RETRY_TIMES = 5;
+    static final String NODE_TOPOLOGY = "/topology";
+
+    private final String subscriptionId;
```

Comment:
I was always confused by this `ZkSubscriptionClient`. For me it's not a client but just a service class that should be stateless. I don't see good reasons for it to hold `subscriptionId` as it can be passed as a method parameter.
@antban @rcillo @lmontrieux @adyach thumbs up my comment if you would like to change this class to be a usual service class or thumbs down if ou want to keep it as it is :)